# DEFINING FUNCTION

In [1]:
def hello():
    print("Hello World")

Here a function named hello is defined which  will print "Hello World".

In [2]:
hello()

Hello World


Here, we are simply calling the function. It will simply call the function definition and the syntax inside the function body is executed which in this case is printing simply "Hello World".

# CALLING FUNCTION WITH PARAMETER

In [3]:
def my_function(name):
    print("Hello " + name)

Here we have defined a function called my-function which takes name as an input parameter and it will print "Hello " + name

In [4]:
my_function("John")

Hello John


As a function is taking one parameter as name so we will pass name = "John".

The output is Hello John

# CALLING FUNCTION WITHOUT PARAMETER

In [5]:
def hello_function():
    print("Hello World, it’s me. Function")      

Here we have not passed any paramter but defined function instead such that when function is being called the statement inside a function is printed only.

In [6]:
hello_function()         

Hello World, it’s me. Function


# DEFAULT PARAMETER VALUE

In [7]:
def my_function(country="India"):
    print("I am from " + country)

Here the parameter is country and its default value is India.So if user passes any value for country then that value is taken as input argument and passed to print statement otherwise the default value which is India is taken into consideration.

In [8]:
my_function("Sweden")

I am from Sweden


Here while calling function we passed Sweden so Sweden will be taken for printing

In [9]:
my_function()

I am from India


Here, we didn't passed anything so bydefault parameter value which is India is printed.

# PYTHON KEYWORD ARGUMENT

In [10]:
def greeting(name, msg = "Good morning!"):  
    print("Hello", name + ', ' + msg)

We have defined a function called greeting which takes two arguments name and msg where msg parameter has default value "Good morning!"

In [11]:
greeting("Bruce")

Hello Bruce, Good morning!


When we only pass Bruce then it is taken as name and msg value is taken as the bydefault value.

In [12]:
greeting("Bruce",msg="How are you ?")

Hello Bruce, How are you ?


Here we define msg as How are you so How are you value for msg is taken for printing and also we define variable name as msg that value "How are you?" is to be allotted to msg only.

In [13]:
greeting(msg="Are you fine?" , name ="john")

Hello john, Are you fine?


Here, we changed the order of arguments as first argument in  function definition  is name but here we are passing "Are you find" but the output is correct as we have defined the variable name that allot "Are you fine?" to msg and "john" to name variable.

In [14]:
greeting(msg="Are you fine?" ,"john")

SyntaxError: positional argument follows keyword argument (<ipython-input-14-149471ec21c5>, line 1)

Here, error is shown as keyword arguments must follow positional arguments.Having a positional argument after keyword arguments will result into errors. 

# NO RETURN STATEMENT

In [15]:
def print_name(name):
       print("Hello your name is " + name)

In [16]:
print_name("Mahesh")

Hello your name is Mahesh


Here we have defined a function with no return type meaning that directly name passed is printed and nothing is returned.

# RETURN STATEMENT

In [17]:
def sum_of_two_numbers(a , b):
       return (a+b)

In [18]:
sum = sum_of_two_numbers(5,6)

Here, we are returning value  of sum in function and result returned is stored into a variable called sum.
The beauty of returning sum is that it can be used in future without again calling function for same value of arguments.

In [19]:
sum

11

# RECURSION

In [20]:
def factorial(n):
    if n==1:
        return 1
    elif n==0:
        return 1
    else:
        return n*factorial(n-1)

In [21]:
factorial(6)

720

In [22]:
factorial(1)

1

In [23]:
factorial(0)

1

Here, instead of making for loop to find factorial of a number we make a function that will call itself until the value of n reaches 1.
When we pass n = 6 then it will proceed to else part and evaluate 6 * fact(5) . As fact(5) is a function so it will again call function fact() with number 5 so now value is 6 * 5 * fact(4) and the loop continues till n=1. As n=1 then it will execute to if condition and 1 is returned.
So final result returned is 6 * 5 * 4 * 3 * 2 * 1 = 720.

# PASSING ONE FUNCTION TO ANOTHER  FUNCTION

In [24]:
def a (): 
       print("Welcome to Function Pass")

def b(function): 
       function()

In [25]:
b(a) 

Welcome to Function Pass


Here, we pass a as a function to function b. As it reaches b then the a function is executed and control will move to function a which will print "Welcome to Function Pass"

# PYTHON LAMBDA (LAMBDA FUNCTION)

In [26]:
a = lambda x,y:x*y

In [27]:
print(a(5,6))

30


Here we have used lamda function which is a one liner and  on passing 5,6 inside a it will evaluate to 5*6 = 30

# MORE ABOUT LAMBDA

In [28]:
def func(n):
    return lambda a:a*n
doubler = func(2)
tripler = func(3)

In [29]:
print(doubler(12))
print(tripler(12))

24
36


Here lambda function can be used inside a function to provide more instruction without complexing the code and also reduces the code length.